src/common/lib/client/baseclient.ts

```diff
@@ -0,0 +1,128 @@
+import Logger, { LoggerOptions } from '../util/logger';
+import Defaults from '../util/defaults';
+import Auth from './auth';
+import ErrorInfo from '../types/errorinfo';
+import Stats from '../types/stats';
+import { StandardCallback } from '../../types/utils';
+import { IHttp, RequestParams } from '../../types/http';
+import ClientOptions, { NormalisedClientOptions } from '../../types/ClientOptions';
+
+import Platform from '../../platform';
+import Rest from './rest';
+import { HttpPaginatedResponse, PaginatedResult } from './paginatedresource';
+
+interface ModulesMap {
+  Rest?: typeof Rest;
+}
+
+class BaseClient {
+  options: NormalisedClientOptions;
+  baseUri: (host: string) => string;
+  authority: (host: string) => string;
+  _currentFallback: null | {
+    host: string;
+    validUntil: number;
+  };
+  serverTimeOffset: number | null;
+  http: IHttp;
+  auth: Auth;
+  _rest?: Rest;
+
+  constructor(options: ClientOptions | string, modules: ModulesMap) {
```

Comment:
```suggestion
  constructor(options: ClientOptions | string, modules?: ModulesMap) {
```

(given the usage without passing modules in `modules.test.js` and the `if (modules?.Rest)` below)